src/vcpkg/portfileprovider.cpp

```diff
@@ -10,10 +11,288 @@
+                {
+                    maybe_scfl =
+                        LocalizedString::from_raw(scfl->control_path)
+                            .append_raw(": ")
+                            .append_raw(ErrorPrefix)
+                            .append(msgMismatchedNames, msg::package_name = port_name, msg::actual = actual_name);
+                }
+            }
+            else
+            {
+                return m_loaded_ports.end();
+            }
+        }
+
+        return m_loaded_ports.emplace_hint(hint, port_name.to_string(), std::move(maybe_scfl));
+    }
+
+    const ExpectedL<SourceControlFileAndLocation>* OverlayPortIndexEntry::try_load_port_subdirectory_with_cache(
+        const ReadOnlyFilesystem& fs, StringView port_name)
+    {
+        if (m_kind != OverlayPortKind::Directory)
+        {
+            Checks::unreachable(VCPKG_LINE_INFO);
+        }
+
+        auto already_loaded = m_loaded_ports.lower_bound(port_name);
+        if (already_loaded == m_loaded_ports.end() || already_loaded->first != port_name)
+        {
+            already_loaded = try_load_port_subdirectory_uncached(already_loaded, fs, port_name);
+        }
+
+        if (already_loaded == m_loaded_ports.end())
+        {
+            return nullptr;
+        }
+
+        return &already_loaded->second;
+    }
+
+    const ExpectedL<SourceControlFileAndLocation>* OverlayPortIndexEntry::try_load_port(const ReadOnlyFilesystem& fs,
+                                                                                        StringView port_name)
+    {
+        switch (determine_kind(fs))
+        {
+            case OverlayPortKind::Port: return try_load_port_cached_port(port_name);
+            case OverlayPortKind::Directory: return try_load_port_subdirectory_with_cache(fs, port_name);
+            case OverlayPortKind::Unknown:
+            default: Checks::unreachable(VCPKG_LINE_INFO);
+        }
+    }
+
+    ExpectedL<Unit> OverlayPortIndexEntry::try_load_all_ports(
+        const ReadOnlyFilesystem& fs, std::map<std::string, const SourceControlFileAndLocation*>& out)
+    {
+        switch (determine_kind(fs))
+        {
+            case OverlayPortKind::Port:
+            {
+                auto& maybe_this_port = *m_loaded_ports.begin();
+                if (auto this_port = maybe_this_port.second.get())
+                {
+                    auto already_in_out = out.lower_bound(maybe_this_port.first);
+                    if (already_in_out == out.end() || already_in_out->first != maybe_this_port.first)
+                    {
+                        out.emplace_hint(already_in_out, maybe_this_port.first, this_port);
+                    }
+
+                    return Unit{};
+                }
+
+                return maybe_this_port.second.error();
+            }
+            case OverlayPortKind::Directory:
+            {
+                auto maybe_subdirectories = fs.try_get_directories_non_recursive(m_directory);
```

Comment:
I considered caching this but decided that the number of times `try_load_all_ports` is called in the same running `vcpkg.exe` is functionally 0 or 1 and thus it wouldn't be worth it.